### DISCUSSION TITLE
Looking for more comprehensive examples using ContainerBuilder<T>

### DISCUSSION THREAD

**Discussion by preetmyob** [testcontainers/testcontainers-dotnet#856] [General]
I've managed to hack my way to getting a MySql container using functional by trial and error and looking at different examples from a number of places.

```csharp
MySqlContainer = new ContainerBuilder<MySqlContainer>()
                .ConfigureDatabaseConfiguration("root", "pw", "db")
                .ConfigureLogging((_, builder) =>
                {
                    builder
                        .AddConsole()
                        .AddDebug();
                })
                .Build();
``` 
Though this allows tests to continue, I still don't see any logging in my tests, so I'm not entirely sure what I set up and where.  

My general point is that most of the examples I'm finding are about a (deprecated) `TestContainersContainer` as opposed to the `ContainerBuilder`. I'd like to see at least some crude example of container startup waiting, for instance. 

Ps; I'm very happy to contribute to new examples if I can btw but might need some pointers on how to do submissions.

**Comment by HofmeisterAn** [comment]
With the latest release `3.0.0`, Testcontainers introduced [modules](https://dotnet.testcontainers.org/modules/) and refactored the API (the docs contains examples of the generic builder too). You are using an obsolete and outdated API. We will remove it as soon as the last legacy module (Azurite) has been migrated. Please use the [MySQL](https://www.nuget.org/packages/Testcontainers.MySql) module:

```console
dotnet add package Testcontainers.MySql
```

```csharp
var mySqlContainer = new MySqlBuilder().Build();
await mySqlContainer.StartAsync()
```

Beside the [docs](https://dotnet.testcontainers.org/) the [tests](https://github.com/testcontainers/testcontainers-dotnet/tree/develop/tests) might also help to understand the module API.

**Reply by preetmyob** [reply]
Thank you @HofmeisterAn

**Reply by preetmyob** [reply]
@HofmeisterAn I've just made the change to use`MySqlBuilder() `instead of `ContainerBuilder`

```csharp
MySqlContainer = new MySqlBuilder().Build();
```

and I'm now getting more errors on startup:

```
Docker.DotNet.DockerApiException
Docker API responded with status code=Conflict, response={"message":"Container b39ea491e83962770ed74a6505527713995f18dbd3418b64d0203639837a6134 is not running"}
```

Looking at the stack trace, I see it's using: `WaitStrategy.WaitUntilAsync`, so I suspect that it's caused by a slower startup. Thus I went about looking for how to use WaitStrategies to increase the timeouts, but I cannot find any examples where they are explicitly usable. 

I did try something like this:

```csharp
MySqlContainer = new MySqlBuilder()
                .WithExposedPort(2343)
                .WithWaitStrategy(
                    WaitStrategy.WaitUntilAsync(
                        new UntilUnixPortIsAvailable(port: 2343),
                        TimeSpan.FromMilliseconds(100),
                        TimeSpan.FromSeconds(20))
                    )
                .Build();
```
All the `Untilxxx` strategies I found were marked `internal`. Would be able to point me at one I would incorporate in my own container starts, please? 


Thank you




---

For ref:

```csharp
   at Docker.DotNet.DockerClient.HandleIfErrorResponseAsync(HttpStatusCode statusCode, HttpResponseMessage response, IEnumerable`1 handlers)
   at Docker.DotNet.DockerClient.MakeRequestAsync(IEnumerable`1 errorHandlers, HttpMethod method, String path, IQueryString queryString, IRequestContent body, IDictionary`2 headers, TimeSpan timeout, CancellationToken token)
   at Docker.DotNet.ExecOperations.ExecCreateContainerAsync(String id, ContainerExecCreateParameters parameters, CancellationToken cancellationToken)
   at DotNet.Testcontainers.Clients.DockerContainerOperations.ExecAsync(String id, IList`1 command, CancellationToken ct)
   at Testcontainers.MySql.MySqlBuilder.WaitUntil.UntilAsync(IContainer container)
   at DotNet.Testcontainers.Containers.DockerContainer.<>c__DisplayClass62_0.<<UnsafeStartAsync>g__CheckWaitStrategy|1>d.MoveNext()
--- End of stack trace from previous location ---
   at DotNet.Testcontainers.Configurations.WaitStrategy.<>c__DisplayClass1_0.<<WaitUntilAsync>g__UntilAsync|0>d.MoveNext()
--- End of stack trace from previous location ---
   at DotNet.Testcontainers.Configurations.WaitStrategy.WaitUntilAsync(Func`1 wait, TimeSpan frequency, TimeSpan timeout, CancellationToken ct)
   at DotNet.Testcontainers.Containers.DockerContainer.UnsafeStartAsync(CancellationToken ct)
   at DotNet.Testcontainers.Containers.DockerContainer.StartAsync(CancellationToken ct)
   at Primitives.Api.Tests.DbWithMigrationsSharedContext.InitializeAsync() in 
```

**Reply by HofmeisterAn** [reply]
> Looking at the stack trace, I see it's using: WaitStrategy.WaitUntilAsync, so I suspect that it's caused by a slower startup.

No, that is not the case. You do not need to change the builder configuration at all. The wait strategy waits until the MySQL database is up, in your case the container just dies (exits) and is not available anymore (while running the readiness checks).

Considering your earlier discussion, I suppose that your Docker environment has not been set up correctly. The following runs without any issues on my machine:

```console
cd tests/Testcontainers.MySql.Tests/
DOCKER_HOST=unix:///Users/atomicjar/.colima/default/docker.sock dotnet test Testcontainers.MySql.Tests.csproj
```

> All the `Untilxxx` strategies I found were marked `internal`.

This is intended, please read the [docs](https://dotnet.testcontainers.org/api/wait_strategies/). As mentioned above you do not need to override the module configuration, `_ = new MySqlBuilder().Build()` should simply work.

**Reply by preetmyob** [reply]
Thank you. I'll investigate that further on my end. It's great that you've shown me how I can run a test against my env. Just to be clear these are intermittent errors that only seem to happen when I run multiple parallel MySQL containers. It does seem that it's likely in my environment, and may well just be resource constraints. I'll do some more testing using other docker environments (windows, Linux directly etc) configured in different ways such as "out of the box" etc. to help me eliminate this. I really appreciate the time you've already spent on my issues, and I greatly appreciate the speed of your responses.

**Reply by HofmeisterAn** [reply]
> Just to be clear these are intermittent errors that only seem to happen when I run multiple parallel MySQL containers.

Probably you run out of memory. My Colima VM (default configuration) has 2GB memory assigned.

**Comment by preetmyob** [comment]
As a follow-up:

1. yes my docker-machine was low on memory (I upped it to 4Gb then 8 Gb) now it works reasonably fast
2. I can use `new MySqlBuilder().Build()....`
3. My tests can do both `mysql:8.0` and `mysql:5.7` images
4. I need to set my `$DOCKER_HOST` env var to ensure the Docker.Net client in test containers was able to connect properly
5. I used a shared context in XUNIT (`CollectionDefinition`) to use just one container for my tests.
6. On my M1 I need to have the `TESTCONTAINERS_RYUK_DISABLED=true` environment variable as the `ryok` cleanup container doesn't seem to work for me - perhaps it's an ARM issue
7. On my Windows x64 machine everything works nicely with `TESTCONTAINERS_RYUK_DISABLED=false` or undefined.

Thank you for all your help @HofmeisterAn

**Reply by HofmeisterAn** [reply]
> 2. I can use `new MySqlContainer()....`

Please use the corresponding builder instead of the container type `new MySqlBuilder().Build()`.

> 6. On my M1 I need to have the TESTCONTAINERS_RYUK_DISABLED=true environment variable as the ryok cleanup container doesn't seem to work for me - perhaps it's an ARM issue

Ryuk runs fine on M1 devices. Disabling Ryuk should not be necessary.

**Reply by preetmyob** [reply]
Apologies I meant to write `new MySqlBuilder().Build()` as that's what I'm using, I really don't know why I stated what I did. Probably a brain fart. 

I've not got ryuk to work yet, so I'll keep plugging away at it.
